src/Analyzer/Resolve/QueryAnalyzer.cpp

```diff
@@ -2919,6 +2919,17 @@ ProjectionNames QueryAnalyzer::resolveFunction(QueryTreeNodePtr & node, Identifi
 
             resolveExpressionNode(in_second_argument, scope, false /*allow_lambda_expression*/, true /*allow_table_expression*/);
         }
+
+        /// Edge case when the first argument of IN is scalar subquery.
+        auto & in_first_argument = function_in_arguments_nodes[0];
+        auto first_argument_type = in_first_argument->getNodeType();
+        if (first_argument_type == QueryTreeNodeType::QUERY || first_argument_type == QueryTreeNodeType::UNION)
+        {
+            IdentifierResolveScope subquery_scope(in_first_argument, &scope /*parent_scope*/);
+            subquery_scope.subquery_depth = scope.subquery_depth + 1;
+
+            evaluateScalarSubqueryIfNeeded(in_first_argument, subquery_scope);
+        }
```

Comment:
It's better to have something like this:
```suggestion
        resolveExpressionNode(in_first_argument, scope, false /*allow_lambda_expression*/, false /*allow_table_expression*/);
```
1. If first argument is an alias to table expression, your solution won't work.
2. `allow_table_expression = false` will treat table expressions as scalar subqueries

Reply:
But you probably can't use table expression in the first argument of IN?
`with (select 1) as t select in(t, 1)` already works, `with t as (select 1) select in(t, 1)` did not work with old analyzer as well (need to check how it works after the change). What would be an example?

I just don't want to re-resolve the first argument. Especially because `resolveExpressionNode` can return a new list of projection names.

Reply:
`with t as (select 1) select in(t, 1)` also works in this PR